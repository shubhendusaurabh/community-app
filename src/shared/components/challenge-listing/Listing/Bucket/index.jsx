/**
 * A single bucket of challenges.
 */

import _ from 'lodash';
import PT from 'prop-types';
import React from 'react';
import Sort from 'utils/challenge-listing/sort';
import SortingSelectBar from 'components/SortingSelectBar';
import Waypoint from 'react-waypoint';
import { getFilterFunction } from 'utils/challenge-listing/filter';
import CardPlaceholder from '../../placeholders/ChallengeCard';
import ChallengeCard from '../../ChallengeCard';
import './style.scss';

const COLLAPSED_SIZE = 10;

export default function Bucket({
  bucket,
  challenges,
  communityName,
  expanded,
  expand,
  loading,
  loadMore,
  openChallengesInNewTabs,
  prizeMode,
  setFilterState,
  setSort,
  sort,
}) {
  const filter = getFilterFunction(bucket.filter);
  const activeSort = sort || bucket.sorts[0];

  const sortedChallenges = _.clone(challenges);
  sortedChallenges.sort(Sort[activeSort].func);

  let expandable = false;
  const filteredChallenges = [];
  for (let i = 0; i < sortedChallenges.length; i += 1) {
    if (filter(sortedChallenges[i])) {
      filteredChallenges.push(sortedChallenges[i]);
    }
    if (!expanded && filteredChallenges.length >= COLLAPSED_SIZE) {
      expandable = true;
      break;
    }
  }

  if (!filteredChallenges.length && !loadMore) return null;

  const cards = filteredChallenges.map(item => (
    <ChallengeCard
      challenge={item}
      newChallengeDetails={!communityName}
      onTechTagClicked={tag => setFilterState({ tags: [tag] })}
      openChallengesInNewTabs={openChallengesInNewTabs}
      prizeMode={prizeMode}
      key={item.id}
    />
  ));

  const placeholders = [];
  if (loading) {
    for (let i = 0; i < 8; i += 1) {
      placeholders.push(<CardPlaceholder id={i} key={i} />);
    }
  }

  return (
    <div styleName="bucket">
      <SortingSelectBar
        onSelect={setSort}
        options={
          bucket.sorts.map(item => ({
            label: Sort[item].name,
            value: item,
          }))
        }
        title={bucket.name}
        value={{
          label: Sort[activeSort].name,
          value: activeSort,
        }}
      />
      {cards}
      {placeholders}
      {
        (expandable || loadMore) && !loading && !expanded ? (
          <button
            onClick={expand}
            styleName="view-more"
          >View more challenges</button>
        ) : null
      }
      {
        !expandable && loadMore && !loading ? (
          <Waypoint onEnter={loadMore} />
        ) : null
      }
    </div>
  );
}

Bucket.defaultProps = {
  communityName: null,
  expanded: false,
  expand: _.noop,
  loading: false,
  loadMore: null,
  openChallengesInNewTabs: false,
  sort: null,
};

Bucket.propTypes = {
  bucket: PT.shape().isRequired,
  expanded: PT.bool,
  expand: PT.func,
  challenges: PT.arrayOf(PT.shape()).isRequired,
  communityName: PT.string,
  loading: PT.bool,
  loadMore: PT.func,
  openChallengesInNewTabs: PT.bool,
  prizeMode: PT.string.isRequired,
  setFilterState: PT.func.isRequired,
  setSort: PT.func.isRequired,
  sort: PT.string,
};
