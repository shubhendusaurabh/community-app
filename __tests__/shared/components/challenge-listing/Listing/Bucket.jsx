import React from 'react';
import _ from 'lodash';
import Renderer from 'react-test-renderer/shallow';
import TU from 'react-dom/test-utils';
import Bucket from 'components/challenge-listing/Listing/Bucket';
import reduxStoreFactory from 'redux-mock-store';
import { Provider } from 'react-redux';

const store = reduxStoreFactory()();

const expand = jest.fn();
const loadMore = jest.fn();
const setFilterState = jest.fn();
const setSort = jest.fn();

const mockDatas = [{
  bucket: {
    sorts: ['current-phase'],
    filter: {
      status: 'ac',
    },
  },
  expanded: true,
  expand,
  challenges: [
    {
      id: '1',
      status: 'b',
      prizes: [1200, 600],
      totalPrize: 1800,
    },
    {
      id: '2',
      status: 'a',
      technologies: 'a, b, c',
      allPhases: [{
        phaseType: 'Registration',
        phaseStatus: 'Open',
      }],
      currentPhases: ['Registration'],
      track: 'DEVELOP',
      subTrack: 'CODE',
      events: [{ eventName: 'Submit' }],
      prizes: [1200, 600],
      totalPrize: 1800,
    }],
  loading: false,
  loadMore,
  setFilterState,
  setSort,
  sort: '',
}, {
  bucket: {
    sorts: ['current-phase'],
    filter: {
      status: 'ac',
    },
  },
  expanded: false,
  expand,
  challenges: [
    {
      id: '1',
      status: 'b',
      prizes: [1200, 600],
      totalPrize: 1800,
    },
    {
      id: '2',
      status: 'a',
      prizes: [1200, 600],
      totalPrize: 1800,
    },
    {
      id: '3',
      status: 'a',
      prizes: [1200, 600],
      totalPrize: 1800,
    },
    {
      id: '4',
      status: 'a',
      prizes: [1200, 600],
      totalPrize: 1800,
    },
    {
      id: '5',
      status: 'a',
      prizes: [1200, 600],
      totalPrize: 1800,
    },
    {
      id: '6',
      status: 'a',
      prizes: [1200, 600],
      totalPrize: 1800,
    },
    {
      id: '7',
      status: 'a',
      prizes: [1200, 600],
      totalPrize: 1800,
    },
    {
      id: '8',
      status: 'a',
      prizes: [1200, 600],
      totalPrize: 1800,
    },
    {
      id: '9',
      status: 'a',
      prizes: [1200, 600],
      totalPrize: 1800,
    },
    {
      id: '10',
      status: 'a',
      prizes: [1200, 600],
      totalPrize: 1800,
    },
    {
      id: '11',
      status: 'a',
      prizes: [1200, 600],
      totalPrize: 1800,
    },
  ],
  loading: true,
  loadMore,
  setFilterState,
  setSort,
  sort: '',
}, {
  bucket: {
    sorts: ['current-phase'],
    filter: {
      status: 'e',
    },
  },
  expanded: false,
  expand,
  challenges: [
  ],
  loading: true,
  setFilterState,
  setSort,
  sort: '',
}];

test('Matches shallow shapshot', () => {
  const renderer = new Renderer();

  _.forEach(mockDatas, (data) => {
    renderer.render((
      <Provider store={store}>
        <Bucket {...data} />
      </Provider>
    ));
    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });
});

class Wrapper extends React.Component {
  componentDidMount() {}

  render() {
    return (
      <Provider store={store}>
        <Bucket {...this.props} />
      </Provider>
    );
  }
}

const instance = TU.renderIntoDocument((<Wrapper {...mockDatas[0]} />));

test('setFilterState', () => {
  const matches = TU.findAllInRenderedTree(instance, item =>
    item && item.className && item.className.match('technology'));
  expect(matches).toHaveLength(3);
  TU.Simulate.click(matches[0]);
  expect(setFilterState).toHaveBeenCalledTimes(1);
  TU.Simulate.click(matches[1]);
  expect(setFilterState).toHaveBeenCalledTimes(2);
  TU.Simulate.click(matches[2]);
  expect(setFilterState).toHaveBeenCalledTimes(3);
});

