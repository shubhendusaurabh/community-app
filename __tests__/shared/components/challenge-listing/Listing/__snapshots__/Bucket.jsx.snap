// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matches shallow shapshot 1`] = `
<Bucket
  bucket={
    Object {
      "filter": Object {
        "status": "ac",
      },
      "sorts": Array [
        "current-phase",
      ],
    }
  }
  challenges={
    Array [
      Object {
        "id": "1",
        "prizes": Array [
          1200,
          600,
        ],
        "status": "b",
        "totalPrize": 1800,
      },
      Object {
        "allPhases": Array [
          Object {
            "phaseStatus": "Open",
            "phaseType": "Registration",
          },
        ],
        "currentPhases": Array [
          "Registration",
        ],
        "events": Array [
          Object {
            "eventName": "Submit",
          },
        ],
        "id": "2",
        "isDataScience": false,
        "prize": Array [
          1200,
          600,
        ],
        "prizes": Array [
          1200,
          600,
        ],
        "status": "a",
        "subTrack": "CODE",
        "technologies": "a, b, c",
        "totalPrize": 1800,
        "track": "DEVELOP",
      },
    ]
  }
  communityName={null}
  expand={[Function]}
  expanded={true}
  loadMore={[Function]}
  loading={false}
  openChallengesInNewTabs={false}
  setFilterState={[Function]}
  setSort={[Function]}
  sort=""
/>
`;

exports[`Matches shallow shapshot 2`] = `
<Bucket
  bucket={
    Object {
      "filter": Object {
        "status": "ac",
      },
      "sorts": Array [
        "current-phase",
      ],
    }
  }
  challenges={
    Array [
      Object {
        "id": "1",
        "prizes": Array [
          1200,
          600,
        ],
        "status": "b",
        "totalPrize": 1800,
      },
      Object {
        "id": "2",
        "prizes": Array [
          1200,
          600,
        ],
        "status": "a",
        "totalPrize": 1800,
      },
      Object {
        "id": "3",
        "prizes": Array [
          1200,
          600,
        ],
        "status": "a",
        "totalPrize": 1800,
      },
      Object {
        "id": "4",
        "prizes": Array [
          1200,
          600,
        ],
        "status": "a",
        "totalPrize": 1800,
      },
      Object {
        "id": "5",
        "prizes": Array [
          1200,
          600,
        ],
        "status": "a",
        "totalPrize": 1800,
      },
      Object {
        "id": "6",
        "prizes": Array [
          1200,
          600,
        ],
        "status": "a",
        "totalPrize": 1800,
      },
      Object {
        "id": "7",
        "prizes": Array [
          1200,
          600,
        ],
        "status": "a",
        "totalPrize": 1800,
      },
      Object {
        "id": "8",
        "prizes": Array [
          1200,
          600,
        ],
        "status": "a",
        "totalPrize": 1800,
      },
      Object {
        "id": "9",
        "prizes": Array [
          1200,
          600,
        ],
        "status": "a",
        "totalPrize": 1800,
      },
      Object {
        "id": "10",
        "prizes": Array [
          1200,
          600,
        ],
        "status": "a",
        "totalPrize": 1800,
      },
      Object {
        "id": "11",
        "prizes": Array [
          1200,
          600,
        ],
        "status": "a",
        "totalPrize": 1800,
      },
    ]
  }
  communityName={null}
  expand={[Function]}
  expanded={false}
  loadMore={[Function]}
  loading={true}
  openChallengesInNewTabs={false}
  setFilterState={[Function]}
  setSort={[Function]}
  sort=""
/>
`;

exports[`Matches shallow shapshot 3`] = `
<Bucket
  bucket={
    Object {
      "filter": Object {
        "status": "e",
      },
      "sorts": Array [
        "current-phase",
      ],
    }
  }
  challenges={Array []}
  communityName={null}
  expand={[Function]}
  expanded={false}
  loadMore={null}
  loading={true}
  openChallengesInNewTabs={false}
  setFilterState={[Function]}
  setSort={[Function]}
  sort=""
/>
`;
